Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorMaskVolumeEffect.py

```diff
@@ -335,31 +358,82 @@ def maskVolumeWithSegment(segmentationNode, segmentID, operationMode, fillValues
+                thresh = vtk.vtkImageThreshold()
+                thresh.SetInputData(inputVolumeNode.GetImageData())
+                thresh.ThresholdByLower(0)
+                thresh.SetInValue(fillValues[1])
+                thresh.SetOutValue(fillValues[1])
+                thresh.SetOutputScalarType(inputVolumeNode.GetImageData().GetScalarType())
+                thresh.Update()
+                stencil.SetInputData(thresh.GetOutput())
+            else:
+                stencil.SetInputData(inputVolumeNode.GetImageData())
+
+            stencil.SetStencilConnection(maskToStencil.GetOutputPort())
+            stencil.SetReverseStencil(operationMode == "FILL_OUTSIDE")
+            stencil.SetBackgroundValue(fillValues[0])
+            stencil.Update()
+            outputVolumeNode.SetAndObserveImageData(stencil.GetOutput())
+        else:
+            # Soft edge
 
-        if operationMode == "FILL_INSIDE_AND_OUTSIDE":
-            # Set input to constant value
             thresh = vtk.vtkImageThreshold()
-            thresh.SetInputData(inputVolumeNode.GetImageData())
+            maskMin = 0
+            maskMax = 255
+            thresh.SetOutputScalarTypeToUnsignedChar()
+            thresh.SetInputData(maskVolumeNode.GetImageData())
             thresh.ThresholdByLower(0)
-            thresh.SetInValue(fillValues[1])
-            thresh.SetOutValue(fillValues[1])
-            thresh.SetOutputScalarType(inputVolumeNode.GetImageData().GetScalarType())
+            thresh.SetInValue(maskMin)
+            thresh.SetOutValue(maskMax)
             thresh.Update()
-            stencil.SetInputData(thresh.GetOutput())
-        else:
-            stencil.SetInputData(inputVolumeNode.GetImageData())
 
-        stencil.SetStencilConnection(maskToStencil.GetOutputPort())
-        stencil.SetReverseStencil(operationMode == "FILL_OUTSIDE")
-        stencil.SetBackgroundValue(fillValues[0])
-        stencil.Update()
+            gaussianFilter = vtk.vtkImageGaussianSmooth()
+            spacing = maskVolumeNode.GetSpacing()
+            standardDeviationPixel = [1.0, 1.0, 1.0]
+            for idx in range(3):
+                standardDeviationPixel[idx] = softEdgeMm / spacing[idx]
+            gaussianFilter.SetInputConnection(thresh.GetOutputPort())
+            gaussianFilter.SetStandardDeviations(*standardDeviationPixel)
+            # Do not truncate the Gaussian kernel at the default 1.5 sigma,
+            # because it would result in edge artifacts.
+            # Larger value results in less edge artifact but increased computation time,
+            # so 3.0 is a good tradeoff.
+            gaussianFilter.SetRadiusFactor(3.0)
+            gaussianFilter.Update()
+
+            import vtk.util.numpy_support
+            maskImage = gaussianFilter.GetOutput()
+            nshape = tuple(reversed(maskImage.GetDimensions()))
+            maskArray = vtk.util.numpy_support.vtk_to_numpy(maskImage.GetPointData().GetScalars()).reshape(nshape)
```

Comment:
I feel like we should have utility method for this, since it shows up a lot.

Reply:
Perhaps `slicer.util.arrayFromVTKImageData` and `slicer.util.arrayFromVTKImageDataModified`

Reply:
Converting a vtkImageData to an ITK-compatible xarray would be indeed nice. However, this would only work well if vtkImageData stored the image geometry (origin, spacing, directions), which we don't do yet in Slicer.

Reply:
No, I just meant that these 4 lines to get a numpy array of a vtkImageData shows up in several places and is kind of ugly.  Implementing xarray would be a different thing, but also good.